Add block hash display case, using middle ellipsis, to ValueCopyable

telestrial opened the pull request:
fixes #451 

Hello! I've been looking at this codebase for a couple of days now and saw a task I thought was approachable for a first PR. This change addresses the issue by modifying the following files:

1. `ExplorerDatum.tsx`
2. `ValueCopyable.tsx`
3. `Block/index.ts`

These are backward-compatible changes, as far as I know, given the optional nature of the props and how the text value in ValueCopyable is calculated.

Issue: `ValueCopyable`, via `ExplorerDatum`, displays the Block hash's first 12 characters, which are typically 0s, with a trailing ellipsis.

Solution: Display the last 12 characters instead, adding the ellipsis at the front. 

How: Add optional maxLength and ellipsisDir props to `ExplorerDatum`, pass them to `ValueCopyable` (adding the optional ellipsisDir prop here), and rewrite the rendered text variable to take advantage of these props, if they exist. When `ellipsisDir="left"` is passed to `ExplorerDatum`, the **last** [maxLength] characters are displayed with an ellipsis in front. When `maxLength={#}` is passed, it overwrites the default maxLength, displaying that many characters before or after the ellipsis. The default ellipsis direction is on the right, as it exists elsewhere today, and the maxLength prop in `ValueCopyable` was already set up well to work without the prop. Now, anywhere `ExplorerDatum` and `ValueCopyable`  or only `ValueCopyable` are used, there are these optional properties to provide more granular display control.

Further considerations (optional): 

1. In `Block/index.ts`, I have only added the ellipsisDir prop set "left". I didn't extend maxLength, but one certainly could, now. I think there are some design questions to ponder in terms of display at < ~500px widths.
2. With the ellipsis on the left, allowing one to display the final characters of very long text, which I'm guessing are more important than the front characters, there aren't a lot of options out there, in terms of a straight CSS solution. Usually, they consist of hacks with non-typical elements. Other javascript solutions could revolve around logic based on `window.clientWidth` or `window.getBoundingClientRect()`, but I've seen some of these solutions and, in my experience, they can be a bit unreliable (the latter function reading 0 for the first milliseconds, for example) and unwieldy in terms of working from hard numerical value breakpoints.
3. One solution to extend the displayed characters could be to flex-col `ExplorerDatum` when it reaches a certain breakpoint (depending on how long you want to go), and, at that point, you *might* want to center `ValueCopyable`. Maybe not, though. It's not something I **explorer**ed.
4. I added this ellipsis left functionality in `Block/index.ts` for both the 'Block hash' and 'Miner payout address' values, only so they would match. You could certainly pull the index out of the map here and apply it only to the block hash.
5. I think the Tooltip component could be useful here, but it would require a more thorough refactor of `ExplorerDatum` which I felt might be beyond the scope of this request. You would probably want to move the logical operator render logic inside the return up above the return, perhaps in a `useCallback`, and then optionally wrap the result of that logic in a Tooltip, if necessary. Figuring out the key for the value you want to display is a barrier.
6. Finally, I added props and imports to the end of things to keep a cleaner diff, but I would probably move maxLength and ellipsisDir next to each other in props and move an import like `useMemo` to the top. Just wanted to keep things simple to review.

Happy to edit anything to your satisfaction. Thanks!

alexfreska reviewed (changes requested):
Hey @telestrial, thank you for submitting a PR!

The approach to the change looks good, but I suggest a few improvements to how the block hash is summarized. I think its important that the user still sees some of the 0s at the start of the hash _and_ a few of the last digits for visual uniqueness. Having both makes it very recognizably a block hash while also providing that uniqueness. 

Here is a how the bitcoin mempool.space explorer does it:
<img width="613" alt="Screenshot 2024-06-13 at 8 21 25 AM" src="https://github.com/SiaFoundation/web/assets/1412796/3656e463-82c5-481d-a739-d6ed0d2480f1">
https://mempool.space/block/0000000000000000000153224d69e68a4bc9027ce2f0833e65cec5ad1f23d579

It will require slightly different props, or a `displayValue` could even be prepared outside the component and passed in - there are a few ways to play with.

alexfreska commented:
In addition, you will need to add a changeset so that this improvement is captured in the next release, please add a changeset file via `npx changeset` with a minor bump to the `explorer` app and a message like:
```
Block hashes now show both leading and ending characters to improve legibility. Closes https://github.com/SiaFoundation/web/issues/451
```

telestrial commented:
Thank you for the feedback! I think this could be done with no added props, then:

`ValueCopyable.tsx`
```typescript
import { useMemo } from 'react'

  const text = useMemo(() => {
    // Display block hashes with an ellipsis in the middle
    if (label?.toLowerCase() === 'block hash') {
      const halfMax = maxLength / 2
      // Floor and ceil here to handle odd maxLengths.
      // .slice() will round down on floats.
      const firstHalf = renderValue.slice(0, Math.floor(halfMax))
      const lastHalf = renderValue.slice(
        renderValue.length - Math.ceil(halfMax)
      )
      return firstHalf + '...' + lastHalf
    }

    return `${renderValue?.slice(0, maxLength)}${
      (renderValue?.length || 0) > maxLength ? '...' : ''
    }`
  }, [label, maxLength, renderValue])
```

The reason I go down to ValueCopyable here is because it's doing its own thing with ellipses. So, even if you do pass in some different value above, it is my (albeit limited) understanding of this code that you're going to need to change ValueCopyable to ignore that in its typical ellipsification. May as well just do the work there, in that case.

If this is a good solution for you, I did run into some trouble with changeset:

The changed package it shows is `@siafoundation/design-system`, and that does make sense to me, given above. But the other issue is that, when I select that package or explorer, it asks me "Which package needs a major bump?" I'm not sure how to/if you can specify something as a minor bump. I think it's calculated automatically..? It may be a peerDependency thing as seen [here](https://github.com/changesets/changesets/blob/main/docs/decisions.md#the-versioning-of-peer-dependencies) under 'The versioning of peer dependencies," but I don't have experience with changeset enough to know for sure.

alexfreska commented:
@telestrial for the changeset CLI, <kbd>Space</kbd> to select, <kbd>Enter</kbd> to move to the next menu. You can always select any unchanged package as show below, and to add a minor or patch change simply press <kbd>Enter</kbd> without a selection to pass the major bump step.

<img width="376" alt="Screenshot 2024-06-14 at 7 39 16 AM" src="https://github.com/SiaFoundation/web/assets/1412796/53d4cbff-e4b5-43e3-90ab-db38a806e49a">

alexfreska reviewed (changes requested):
@telestrial this looks good except for the ` if (label?.toLowerCase() === 'block hash') {` check. We should let `label` be anything and not use it to determine the type of data, `label` could be "block hash", or "previous block hash" for a different feature, or translated i18n value in the future. The ValueCopyable component supports passing an `type` prop (`EntityType`) which includes `block`, but this is currently for a block number, eg 149,000, not the hash.

I think the correct approach here would be adding a `blockHash` type to `EntityType` and then format `renderValue` based on the `type` in the same way its used to grab a default label and display length. In addition, the `displayValue` should always override this internal behaviour.
```diff
  const label = customLabel || getEntityTypeCopyLabel(type)
  const maxLength = customMaxLength || getEntityDisplayLength(type)
  const cleanValue = stripPrefix(value)
-  const renderValue = displayValue || cleanValue
+  const renderValue = displayValue || formatEntityValue(type, cleanValue)
```

telestrial commented:
I completely agree. I went down this road when I first looked at this, before realizing the data wasn't being typed like that. I didn't want to make such a big change without knowing more, and I'm glad I didn't. I would have assumed the block entitytype fit. Happy to do this, extending the type, adding it at `Block/index.tsx`, passing it down, and coming up with a formatting function. Is there a place where you're keeping visual helper functions like the one proposed? `/lib`?

telestrial commented:
Third time's the charm! And, if not, no worries: between your comments and my time spent working on the project, I am learning so much about the codebase on each round of this. It's fun!

So the goal was to get here, as you stated:

```typescript
const renderValue = displayValue || formatEntityValue(type, cleanValue)
```

Here is where I landed with that:

`ValueCopyable.tsx`
```typescript
  // If we pass a defaultValue, use it. An entityType? Format it.
  const renderValue =
    displayValue || (type && formatEntityValue(type, cleanValue, maxLength))
  // If we don't yet have a renderValue, apply some general formatting.
  const text = renderValue || defaultFormatValue(cleanValue, maxLength)
```

This creates the `displayValue > entityValue > value` priority.

Additionally:

* Extended `EntityTypes` for 'blockHash' and wrote two functions: `formatEntityValue` and `defaultFormatValue`.
* In `Block/index.tsx`, formatted blockHash correctly *and* added the transaction entityType down below.  Everything was set up below in EntityList to treat these transactions as their entity type, but it wasn't typed that way up top.
* In `ExplorerDatum.tsx`, added and passed the displayValue prop and blocked an edge case with getHref where it expects blockHash to have a value in `routes.ts`. That's error-boundary/500 territory.

I checked Contract and Host, and I believe now that most entity data is being taken through formatEntityValue and most text through value, but, having written most/all of this, you might want to do some checking there. I think that should be the ultimate goal--separating these paths so that you can modify the treatment of one without affecting the other. I think there may be some cases where you are passing entity data as value because you want a specific format, but that may now work best on the displayValue prop, which didn't exist before. I didn't want to pre-suppose your intentions in those places, though.

Let me know how this lands. Happy to continue to work on this until it matches your vision.

alexfreska commented on apps/explorer/components/ExplorerDatum.tsx:
```diff
@@ -48,13 +50,16 @@ export function ExplorerDatum({
             <ValueCopyable
               scaleSize="18"
               label={entityType}
```
Looks like an existing bug getting surfaced, but I noticed this label being set to `label={entityType}` is causing the custom label to override `getEntityTypeCopyLabel`, I think we can safely remove this line and fix that issue.

<img width="516" alt="Screenshot 2024-06-18 at 10 48 22 AM" src="https://github.com/SiaFoundation/web/assets/1412796/e486b065-3abe-4946-a350-04d863c785a5">

telestrial replied:
Good catch! Got it.

alexfreska reviewed (changes requested):
This looks great! Added two last small change suggestions, after those lets merge this. As you can see this component is probably overloaded with too many options tied to implicit combinations of props, in the future I think it could be cleaned up into a component that delegates out to more specific components with clearer logic.

telestrial commented:
Thanks! And yeah, I did notice that. I like the delegation idea, sort of in the same vein as `ExplorerDatum`, pitching to different components below it, but another approach could be from above ValueCopyable instead, treating it as the base component. You can't actually write anything like this, but the idea would be the same vein as one type implementing another: `EntityValueCopyable implements ValueCopyable`. Wrappers. I think this would push previous and future code in that desirable direction--fewer optional props, separation of concerns, insulation of one project/type from any specialization another may need, etc. Less room to introduce problems, in general, because each component would have less on its plate. Worlds more extensible for whatever the future may bring. Downside: more components to inventory. It's probably worth it.

I'll get these small changes in here tonight or tomorrow and let you know!

telestrial commented on apps/explorer/components/ExplorerDatum.tsx:
```diff
@@ -48,13 +50,16 @@ export function ExplorerDatum({
-                entityType === 'block' && entityValue
-                  ? Number(entityValue).toLocaleString()
-                  : entityValue
+              href={
```
Got this covered. One thought for the future: it may be preferable from a UX perspective to rewrite the URL on Block to either always use the block or always use the blockHash as the id. Could also redirect, but that's a bit heavier than needed, maybe. I'd personally want the functionality to be such that if you clicked the link for the "other" route going to this same page, nothing would happen. Would need to look into the right strategy for that. It's a bit of a feeling, but it just seems like a different ID on the same base route like this should go somewhere else or it just shouldn't be a visible thing. Internally, the functionality of accepting both is excellent.

alexfreska replied:
Yeah true, thats a change probably worth exploring.

alexfreska reviewed (approved):
Looks great! Running CI checks now. Looks like you need to remove the `.` from the end of one of the commit messages. 
Also, please squash the commits into 1 commit encapsulating the feature/change, you can do this locally or with the merge option on GitHub!

<img width="791" alt="Screenshot 2024-06-20 at 9 19 26 AM" src="https://github.com/SiaFoundation/web/assets/1412796/ecfae97f-581b-40e8-bc1e-f2a1b1a145a6">

alexfreska commented:
Merging, thank you for contributing! ✨ 

Note: The relevant lint/builds/tests are passing, some unrelated test runs do not have the correct permissions to run on the forked branch, they are currently be reconfigured to support this.

telestrial commented:
Awesome. Thank you for working with me on it! Your guidance here was paramount. I very much appreciate it.
